Allow dot'ted field in connectFrom of $graphLookup aggregation pipeline

arunl opened the pull request:
This relates to Issue #595. This adds capability for using dot'ed parameters in connectionFrom field of $graph aggregation pipeline.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mongomock/mongomock/686)
<!-- Reviewable:end -->

arunl committed:
Allow dot'ted connectFrom in $graphLookup

Modified the original code. Split the connectFrom parameters, and
recursively traverse the split list and the document (in parallel).
If the value of a parameter is a list, it traverses all the elements
of the list. If a dictionary, it traverses values of the
matching key.

arunl committed:
Removed test expected to fail with dot'ted connectFrom parameter

arunl committed:
Added unit tests for dot'ed connectFrom $graphlookup

arunl committed:
Ignore *~ emacs temporary files

arunl committed:
Mostly reformatting

arunl committed:
Made pep8 and pylint compliant

arunl committed:
Reorganized code to address pylint and pep8 issues

arunl committed:
More fixes for pylint compliance

Removed parameterized test using pytest. Replaced with unittest.
I am guessing unittest is preferred for compatibility with other
versions and operating environments.

arunl committed:
Replaced yield from by for-in/yield

Why? For python 2.7 compatibility

arunl committed:
Missed a yield-from earlier; removed it

arunl committed:
Made tests nose-compliant

arunl committed:
Oops! Removed duplicate function that was source of yield from problem

The function '_recursive_get' had yield-from, which is not supported by
python 2.7. I thought I had changed it to an iterative yield. Turns out
I had duplicate copy of this function. I had been changing one, and the
other one was triggering Travis CI failure. I hope this settles the
issue.

arunl committed:
Using tests.diff to compare actual and expected

I was earlier using assertEqual(actual, expected), which was good enough
for my work. I expected it would break somepoint. And it broke the CI.

Looking around for options I lucked into the 'test__diff.py', and
learned about 'tests.diff'. Am using that now instead.

arunl committed:
Changes for flake8 compliance

arunl committed:
Ordered imports

arunl committed:
Trying to get past flake8 H405 violation

Am flummoxed by this violation. Cannot recreate it on my machine. flake8
ran independently gives no error. tox -e pep8 (with python 3.6.7 and
3.8) errors out, with an exception that FlakeChecker doesn't have an
  attribute CONSTANT.

arunl committed:
Last fixup for H405 error

Required a lot of work for, in my opinion, a very superfluous coding
standard. To beat it I changed the docstring to be a single line
comment. It made the comment less readable, but it got me past H405.

The bright side, I was able to recreate the issue on my machine. It
required making a change to pyflakes/checker.py. For anyone experiencing
the same problem, here is the fix I used:

In pyflakes 1.2.3, file pyflakes/checker.py, change line 854

from:

        STARRED = NAMECONSTANT = handleChildren
to:
        STARRED = NAMECONSTANT = CONSTANT = handleChildren

arunl commented:
Am pretty close to getting thru the pipeline. The code coverage is failing because my tests are not covering one additional statement, compared to the original. I could use help with which statement am I missing.  I am not able to get the details from codecove report. (It just could be that I connected my repo after this failure. Will give it a try after making some dummy changes.)

arunl committed:
Change mostly to rerun coverage

pcorpet commented on .gitignore:
```diff
@@ -51,3 +51,4 @@ flycheck-*
 
 ### vim
 .ropeproject/
+*~
```
Please don't modify this file, instead update your local .gitignore to cover for those files.

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -952,7 +952,7 @@ def _handle_graph_lookup_stage(in_collection, database, options):
         if options[operator].startswith('$'):
             raise OperationFailure("FieldPath field names may not start with '$'")
-        if operator in ('connectFromField', 'as') and \
+        if operator in ('as') and \
```
Thanks for implementing, please simplify the condition here (`==` instead of `in`) and make it fit on one line.

Note that `operator in ('as')` would return true only for operator being`a` or `s`, not `as` (that might be the missing line in coverage).

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
+        else:
+            # more to go, get remaining from list/dict
+            remaining_fields = nested_fields[1:]
+            if isinstance(head, list):
```
Please check for dict or tuples: `if isinstance(head, (list, tuple)):`

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
+        if len(nested_fields) == 1:
+            # final/last field reached
+            yield head
+        else:
```
I suggest you return early in the branch above instead of adding the whole code in an `else` branch:

```py
if len(nested_fields) == 1:
  yield head
  return

# More to go: get remaining from list/dict.

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
+        first_field = nested_fields[0]
+        head = match.get(first_field)
+        if len(nested_fields) == 1:
+            # final/last field reached
```
Nit - Please start your comments with an upper case letter and end them with a dot.

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
                 found_items.add(new_match['_id'])
         return new_matches
 
+    def _recursive_get(match, nested_fields):
```
This function doesn't use anything apart from its parameters, so please move it at the module level.

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
+            yield head
+        else:
+            # more to go, get remaining from list/dict
+            remaining_fields = nested_fields[1:]
```
For better readability I suggest that you move this before line 991 and check remaining_fields instead:

```py
head = match.get(nested_fields[0])
remaining_fields = nested_fields[1:]
if not remaining_fields:
  yield head
```

pcorpet commented on mongomock/aggregate.py:
```diff
@@ -984,6 +984,28 @@ def _find_matches_for_depth(query):
+                    yield answer
+            else:
+                # field doesn't exist, just let go
+                pass
```
I'm not sure we need this code: dropping the else and pass statement should work the same.

pcorpet commented on tests/test__collection_api.py:
```diff
@@ -3153,21 +3153,6 @@ def test__aggregate_graph_lookup_depth_filed_not_string(self):
             "Argument 'depthField' to $graphlookup must be a string",
             str(err.exception))
 
-    def test__aggregate_graph_lookup_dot_in_connect_from_field(self):
```
Please add a test for dotted field in `as` if there isn't one (it should trigger a NotImplementedError)

pcorpet reviewed (changes requested):
Thanks a lot for implementing that feature.

I've noted that you're introducing a new architecture for test files. I would prefer that we do that in a separate PR. For now please try to fit in the existing test framework and maybe add a TODO to split it in smaller test modules.

Also I would very much like to add tests in test__mongomock that make sure that the behavior of mongomock stays the same as pymongo + MongoDB.

arunl committed:
Added a failure test for dot in 'as' field

arunl committed:
Moved *~ to local gitignore

arunl commented:
@pcorpet  Re: changing test architecture.

Like to make sure I have understood you correctly. You are asking me to do the following:

1) Move the tests I have added into test__collection_api.py. These tests do not use live mongo server. 
2) Add more tests in test test__mongomock that compares the behavior of mongomock with pymongo + MongoDB.

Did I get you right?

arunl committed:
Changed as per suggestions by @pcorpet

- Moved test cases into test__collection.py
- Refactored aggregate.py for better code

arunl committed:
More pep8 fixes not caught locally by flake8/pycodestyle

arunl committed:
Hopefully this is the last PEP8 issue

arunl committed:
Playing whackamole with PEP8 compliance

arunl commented:
@pcorpet I've addressed all your comments, except one, which happens to be the most significant one, in my opinion. The pending task is to add tests in test__mongomock. 

I have reviewed how you have written other tests. I think it will be simple enough to translate the tests I have in "test__collections_api" into the dual comparison tests.  But it will likely take some time. I will need to switch context to my day job.

pcorpet commented on .gitignore:
```diff
@@ -51,3 +51,4 @@ flycheck-*
 
 ### vim
 .ropeproject/
+
```
Please drop that line as well so that there is no change in this file for your PR.

pcorpet reviewed (changes requested):
Thanks for the changes. This seems OK to me. Just add the same tests in test__mongomock (without the expected results, as the expectations should just be to match the pymongo + MongoDB behavior).

arunl committed:
Added graphlookup comparison tests between real and mock mongo

arunl commented:
@pcorpet - I added tests in test__mongomock. Was very easy. You've set up the testing architecture very well.

pcorpet commented on tests/test__mongomock.py:
```diff
@@ -3571,6 +3571,118 @@ def test_aggregate_project_with_boolean(self):
+
+        self.cmp_a.do.insert_many(data_a)
+        self.cmp_b.do.insert_many(data_b)
+        self.cmp_b.do.aggregate(query)
```
Please use self.cmp_b.compare, same in the other tests

pcorpet committed:
Compare results in new tests in test__mongomock.

pcorpet commented:
Thank you for the hard work. Feel free to send another PR to add your name in the contributors of the README (I should have suggested that earlier, sorry).
